### PR TITLE
Stream docs

### PR DESCRIPTION
**Pull request by AlexanderBenz**

### CODE REVIEW

File: docs/examples/stream/_density_uncertainty.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
+[
+  {
+    "class" : "StreamDensityBasedAL",
+    "package" : "stream",
+    "method" : "StreamDensityBasedAL Sampling",
+    "category" : "Density Sampling",
+    "template" : "examples/template_stream.py",
+    "tags" : ["stream",  "classification", "single-annotator"],
+    "title" : "Density Sampling with StreamDensityBasedAL",
+    "refs" : ["settles2009active"],
```
Comment (tpham93) [inline docs/examples/stream/_density_uncertainty.json:10]: Fix reference: https://proceedings.mlr.press/v36/ienco14.html

File: docs/examples/stream/_density_uncertainty.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  },
+  {
+    "class" : "CognitiveDualQueryStrategyRan",
+    "package" : "stream",
+    "method" : "CognitiveDualQueryStrategyRan Uncertainty Sampling",
+    "category" : "Density Sampling",
+    "template" : "examples/template_stream.py",
+    "tags" : ["stream",  "classification", "single-annotator"],
+    "title" : "Density Sampling with CognitiveDualQueryStrategyRan",
+    "refs" : ["settles2009active"],
```
Comment (tpham93) [inline docs/examples/stream/_density_uncertainty.json:26]: Fixe Reference for all CogDQS strategies: https://ieeexplore.ieee.org/document/9492291/

File: docs/examples/stream/_stream_baselines.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
+[
+  {
+    "class" : "StreamRandomSampling",
+    "package" : "stream",
+    "method" : "StreamRandomSampling Sampling",
+    "category" : "Random Sampling",
+    "template" : "examples/template_stream.py",
+    "tags" : ["stream",  "classification", "single-annotator"],
+    "title" : "Sampling with StreamRandomSampling",
+    "refs" : ["settles2009active"],
```
Comment (tpham93) [inline docs/examples/stream/_stream_baselines.json:10]: Remove reference

File: docs/examples/stream/_stream_baselines.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  },
+  {
+    "class" : "PeriodicSampling",
+    "package" : "stream",
+    "method" : "PeriodicSampling Sampling",
+    "category" : "Random Sampling",
+    "template" : "examples/template_stream.py",
+    "tags" : ["stream",  "classification", "single-annotator"],
+    "title" : "Sampling with PeriodicSampling",
+    "refs" : ["settles2009active"],
```
Comment (tpham93) [inline docs/examples/stream/_stream_baselines.json:27]: Remove reference

File: docs/examples/stream/_stream_probabilistic_al.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
+[
+  {
+    "class" : "StreamProbabilisticAL",
+    "package" : "stream",
+    "method" : "StreamProbabilisticAL Sampling",
+    "category" : "Probalistic Sampling",
+    "template" : "examples/template_stream.py",
+    "tags" : ["stream",  "classification", "single-annotator"],
+    "title" : "Sampling with StreamProbabilisticAL",
+    "refs" : ["settles2009active"],
```
Comment (tpham93) [inline docs/examples/stream/_stream_probabilistic_al.json:10]: https://link.springer.com/chapter/10.1007/978-3-319-24465-5_13

File: docs/examples/stream/_uncertainty_sampling.json
Comment (tpham93) [inline docs/examples/stream/_uncertainty_sampling.json:1]: Fix references: https://ieeexplore.ieee.org/document/6414645

File: docs/examples/template_stream.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+X_train.extend(X_init)
+y_train = []
+y_train.extend(y_init)
+queried_count = 0
+budget_list = []
+
+t_ax = [0] * len(y_init)
+acq = [True] * len(y_init)
+# colors = []
+# mesh_instances = np.array([X_mesh.reshape(-1), Y_mesh.reshape(-1)]).T
```
Comment (tpham93) [inline docs/examples/template_stream.py:65]: cleanup

File: docs/examples/template_stream.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+for t_x, (x_t, y_t) in enumerate(zip(X_stream, y_stream)):
+
+    X_cand = x_t.reshape([1, -1])
+    y_cand = y_t
+    clf.fit(X_train, y_train)
+    # Get labeled instances.
+    X_labeled = np.array(X_train)[labeled_indices(y_train)]
+    # check whether to sample the instance or not
+    X_train_array = np.array(X_train).reshape([-1, 1])
```
Comment (tpham93) [inline docs/examples/template_stream.py:77]: X_labeled and X_train_array (74-77) not used anywhere

File: docs/examples/template_stream.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    # Label the queried instances.
+    X_train.append(x_t)
+    if len(sampled_indices):
+        y_train.append(y_t)
+    else:
+        y_train.append(MISSING_LABEL)
+
+    t_ax.append(t_x)
+    acq.append(len(sampled_indices) > 0)
+    # colors.append("r" if y_cand == classes[0] else "b")
```
Comment (tpham93) [inline docs/examples/template_stream.py:93]: cleanup

File: docs/examples/template_stream.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            ax, t_x, plot_step, clf, X, predictions_list
+        )
+        data_lines = plot_stream_training_data(
+            ax, t_ax, X_train, acq, y_train, classes, feature_bound
+        )
+
+        title_string = (
+            f"Decision boundary after {t_x} new instances \n"
+            + f"with utility: {utilities[0]: .4f} "
+            + f"budget is {budget_list[-1]:.4f}"
```
Comment (tpham93) [inline docs/examples/template_stream.py:111]: Replace by queried_count/(tx+1) and remove budget_list

File: skactiveml/stream/_density_uncertainty.py [hunk 10]
```diff
@@ -157,13 +157,14 @@ def query(
             predict_proba[:, utilities_index[:, 1]]
             - predict_proba[:, utilities_index[:, 0]]
         )
+        utilities = 1 - utilities[0]
```
Comment (tpham93) [inline skactiveml/stream/_density_uncertainty.py:160]: rename in confidence

File: skactiveml/stream/_density_uncertainty.py [hunk 11]
```diff
@@ -622,6 +623,7 @@ def query(
         # its the margin but used as utillities
         predict_proba = clf.predict_proba(candidates)
         utilities = np.max(predict_proba, axis=1)
+        utilities = 1 - utilities
```
Comment (tpham93) [inline skactiveml/stream/_density_uncertainty.py:626]: rename in confidence

File: skactiveml/stream/_uncertainty_zliobaite.py [hunk 12]
```diff
@@ -130,6 +130,7 @@ def query(
 
         predict_proba = clf.predict_proba(candidates)
         utilities = np.max(predict_proba, axis=1)
+        utilities = 1 - utilities
```
Comment (tpham93) [inline skactiveml/stream/_uncertainty_zliobaite.py:133]: rename in confidence

File: skactiveml/visualization/tests/test_feature_space.py
Comment (tpham93) [inline skactiveml/visualization/tests/test_feature_space.py:1]: Add tests where pdf files are compared so look of plots is consistent. See pool tests where compare_images is used

**Review by tpham93 — CHANGES_REQUESTED** [review]

File: docs/examples/template_stream.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+queried_indices = [True] * len(y_init)
+
+predictions_list = []
+
+for t_x, (x_t, y_t) in enumerate(zip(X_stream, y_stream)):
+
+    X_cand = x_t.reshape([1, -1])
+    y_cand = y_t
+    clf.fit(X_train, y_train)
+    # check whether to sample the instance or not
```
Comment (tpham93) [inline docs/examples/template_stream.py:69]: Capitalisation

**Review by tpham93 — COMMENTED** [review]
Looks good now.

File: docs/examples/template_stream.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+"$import_clf|from skactiveml.classifier import ParzenWindowClassifier"
+"$import_misc"
+
+random_state = np.random.RandomState(0)
+init_size = "$init_size|0"
+# Build a dataset.
+X, y_true = make_blobs(
+    n_samples="$n_samples|200" + init_size,
+    n_features=2,
```
Comment (mherde) [inline docs/examples/template_stream.py:19]:
```suggestion
    n_features=1,
```

File: docs/examples/template_stream.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    centers=[[0], [-3], [1], [2], [-0.5]],
+    cluster_std=0.7,
+    random_state=random_state,
+)
+y_true = y_true % 2
+X_init = X[:init_size, :]
```
Comment (mherde) [inline docs/examples/template_stream.py:25]:
```suggestion
X_init = X[:init_size]
```

File: docs/examples/template_stream.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+y_init = y_true[:init_size]
+X_stream = X[init_size:, :]
```
Comment (mherde) [inline docs/examples/template_stream.py:27]:
```suggestion
X_stream = X[init_size:]
```

File: docs/examples/template_stream.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+            ax, t_x, plot_step, clf, X, predictions_list, res="$res|25"
+        )
+        data_lines = plot_stream_training_data(
+            ax, X_train, y_train, queried_indices, classes, feature_bound
+        )
+
+        title_string = (
+            f"Decision boundary after {t_x} new instances \n"
+            + f"with utility: {utilities[0]: .4f} "
+            + f"budget is {queried_count / (t_x + 1):.4f}"
```
Comment (mherde) [inline docs/examples/template_stream.py:100]:
```suggestion
            + f"budget is {sum(queried_indices) / (t_x + 1):.4f}"
```

File: docs/examples/template_stream.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    qs.update("$update_params")
+    # Label the queried instances.
+    X_train.append(x_t)
+    if len(sampled_indices):
+        y_train.append(y_t)
+    else:
+        y_train.append(MISSING_LABEL)
+
+    queried_indices.append(len(sampled_indices) > 0)
+    queried_count += len(sampled_indices)
```
Comment (mherde) [inline docs/examples/template_stream.py:84]:
```suggestion
```

File: skactiveml/stream/_density_uncertainty.py [hunk 19]
```diff
@@ -153,17 +153,18 @@ def query(
         # calculate the margin used as utillities
         predict_proba = clf.predict_proba(candidates)
         utilities_index = np.argpartition(predict_proba, -2)[:, -2:]
-        utilities = (
+        confidence = (
             predict_proba[:, utilities_index[:, 1]]
             - predict_proba[:, utilities_index[:, 0]]
         )
+        utilities = 1 - confidence[0]
```
Comment (mherde) [inline skactiveml/stream/_density_uncertainty.py:160]: Check indexing.

File: skactiveml/visualization/_feature_space.py [hunk 20]
```diff
@@ -345,6 +345,157 @@ def plot_contour_for_samples(
...
+    plot_cand_highlight=True,
+):
+    """Plot the utility for the given query strategy.
+
+    Parameters
+    ----------
+    ax : matplotlib.axes.Axes
+        The axis on which the utility is plotted. Only if y.ndim = 1 (single
+        annotator).
+    X : array-like of shape (n_samples, n_features)
```
Comment (mherde) [inline skactiveml/visualization/_feature_space.py:368]:
```suggestion
    X : array-like of shape (n_samples, 1)
```

File: skactiveml/visualization/_feature_space.py [hunk 21]
```diff
@@ -345,6 +345,157 @@ def plot_contour_for_samples(
...
+        Training data set, usually complete, i.e. including the labeled and
+        unlabeled samples.
+    y : array-like of shape (n_samples, )
+        Labels of the training data set (possibly including unlabeled ones
+        indicated by self.MISSING_LABEL).
+    queried_indices : array-like of shape (n_samples,)
+        Indicates which instances from candidates have been queried.
```
Comment (mherde) [inline skactiveml/visualization/_feature_space.py:375]:
```suggestion
        Indicates which instances from candidates have been queried.
    classes : ...
```

File: skactiveml/visualization/_feature_space.py [hunk 22]
```diff
@@ -345,6 +345,157 @@ def plot_contour_for_samples(
...
+        Set the line width in points.
+    plot_cand_highlight: bool
+        The indicator to higlight the current candidate.
+
+    Returns
+    -------
+     axes : array-like of shape (n_annotators_to_plot,)
+         The axes on which the utilities were plotted.
+    """
+    check_array(X, ensure_2d=False)
```
Comment (mherde) [inline skactiveml/visualization/_feature_space.py:397]: check shape of `X` to be `(n_samples, 1)`

File: skactiveml/visualization/_feature_space.py [hunk 23]
```diff
@@ -345,6 +345,157 @@ def plot_contour_for_samples(
...
+    ax: matplotlib.axes.Axes or List
+        The axis on which the decision boundary is plotted. If ax is a List,
+        each entry has to be an `matplotlib.axes.Axes`.
+    t_x: int
+        The position of the newest instance for the x axies.
+    plot_step: int
+        The interval in which the clf should predict new samples.
+    clf: Sklearn classifier
+        The fitted classifier whose decision boundary is plotted.
+    X : array-like of shape (n_samples, n_features)
```
Comment (mherde) [inline skactiveml/visualization/_feature_space.py:459]:
```suggestion
    X : array-like of shape (n_samples,)
```

File: skactiveml/visualization/_feature_space.py [hunk 24]
```diff
@@ -345,6 +345,157 @@ def plot_contour_for_samples(
...
+        Training data set, usually complete, i.e. including the labeled and
+        unlabeled samples.
+    pred_list: array-like of shape (n_samples, )
+        The list containing classifier prediction for the last steps.
+    color: str | matplotlib.colors.Colormap, optional (default='k')
+        The color for the decision boundary.
+    res : int, optional (default=25)
+        The resolution of the plot.
+    Returns
```
Comment (mherde) [inline skactiveml/visualization/_feature_space.py:468]:
```suggestion

    Returns
```

File: skactiveml/visualization/_feature_space.py [hunk 25]
```diff
@@ -345,6 +345,157 @@ def plot_contour_for_samples(
...
+    -------
+    ax: matplotlib.axes.Axes or List
+        The axis on which the boundary was plotted or the list of axis if ax
+        was a list.
+    pred_list: array-like of shape (n_samples, )
+        The list containing classifier prediction for the last steps.
+    """
+    X = check_array(X, ensure_2d=False)
+    check_array(pred_list, ensure_2d=False, ensure_min_samples=0)
```
Comment (mherde) [inline skactiveml/visualization/_feature_space.py:477]: Check whether it needs to be a column array, then use `check_column` instead.

**Review by mherde — CHANGES_REQUESTED** [review]

**Review by tpham93 — APPROVED** [review]

**Review by mherde — APPROVED** [review]

**Pull request merged**